File: src/structures/Document.java
package structures;

import java.io.Serializable;

/**
 * Created by wolf on 08.07.2015.
 */
public class Document implements Serializable {
    public static String TWEET = "T";
    public static String BLOG = "B";
    public static String REUTORS = "R";
    public static String PAN = "P";
    public static String ARTICLE = "A";
    public static String UNKNOWN = "UNKNOWN";

    protected String type;
    protected String author;
    protected String text;
    protected String docid;

    public Document() {
    }

    public Document(String author, String text) {
        this.author = author;
        this.text = text;
    }

    public Document(String docid, String type,String text) {
        this.docid = docid;
        this.text = text;
        this.type = type;
        this.author = UNKNOWN;
    }

    public Document(String docid, String type,String text,  String author) {
        this.docid = docid;
        this.text = text;
        this.type = type;
        this.author = author;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public boolean isEmpty(){
        return author==null || text==null || author.isEmpty() || text.isEmpty();
    }

    public boolean isBlog(){
        return type!=null && type.equals(BLOG);
    }

    public boolean isArticle(){
        return type!=null && type.equals(ARTICLE);
    }

    public boolean isReuters(){
        return type!=null && type.equals(REUTORS);
    }

    public boolean isBlogArticle(){
        return type!=null && (type.equals(ARTICLE) || type.equals(BLOG));
    }

    public boolean isTweet(){
        return type!=null && type.equals(TWEET);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Document document = (Document) o;

        if (type != null ? !type.equals(document.type) : document.type != null) return false;
        return !(author != null ? !author.equals(document.author) : document.author != null);

    }

    @Override
    public int hashCode() {
        int result = type != null ? type.hashCode() : 0;
        result = 31 * result + (author != null ? author.hashCode() : 0);
        return result;
    }
}

File: src/data/document/Document.scala
package data.document

import data.dataset.XMLParser
import language.boundary.SentenceML
import language.morphology.MorphTagger
import language.postagger.PosTaggerImp
import language.tokenization.TokenizerImp
import options.Resources
import options.Resources.DocumentModel

import scala.collection.JavaConversions

/**
  * Created by wolf on 01.04.2016.
  */
case class Document(var docid: String, var doctype: String, var author: String) extends Serializable {
  var genre: String = null
  var title: String = null
  var text: String = null

  def this(docid: String, doctype: String) = this(docid, doctype, null)

  var paragraphs = Seq[String]()

  ///////////////////////////////////////////////////////////
  //////////////////////////////////////////////////////////
  //<editor-fold defaultstate="collapsed" desc="Setters and Getters">
  def setGenre(genre: String): this.type = {
    this.genre = genre
    this
  }

  def setTitle(title: String): this.type = {
    this.title = title
    this
  }

  def setText(text: String): this.type = {
    this.text = text
    this
  }

  def setParagraphs(paragraphs: Seq[String]): this.type = {
    this.paragraphs = paragraphs
    this
  }

  def setParagraphs(paragraphs: java.util.List[String]): this.type = {
    this.paragraphs = JavaConversions.asScalaBuffer(paragraphs).toSeq
    this
  }

  def getGenre(): String = {
    genre
  }

  def getTitle(): String = {
    title
  }

  def getText(): String = {
    text
  }

  def getParagraphs(): Seq[String] = {
    paragraphs
  }


  def getAuthor(): String = {
    author
  }

  def getDocId(): String = {
    docid
  }

  def getDocType(): String = {
    doctype
  }


  //</editor-fold>
  ///////////////////////////////////////////////////////////
  //////////////////////////////////////////////////////////

  ///////////////////////////////////////////////////////////
  //////////////////////////////////////////////////////////
  //<editor-fold defaultstate="collapsed" desc="is methods">
  def isBlog(): Boolean = {
    DocumentModel.isBlog(doctype)
  }

  def isTweet(): Boolean = {
    DocumentModel.isTweet(doctype)
  }

  def isArticle(): Boolean = {
    DocumentModel.isArticle(doctype)
  }

  def isPAN(): Boolean = {
    DocumentModel.isPAN(doctype)
  }

  def hasLengthGreater(length: Int): Boolean = {
    text.length > length
  }

  def toParagraphs(): Seq[Paragraph] = {
    if (!paragraphs.isEmpty) {
      paragraphs.zipWithIndex.map { case (text, seq) => {
        new Paragraph(docid, author, seq, text)
      }
      }
    }
    else {
      //Paragraphize here
      Seq(new Paragraph(docid, author, 0, text))
    }
  }


  //</editor-fold>
  ///////////////////////////////////////////////////////////
  //////////////////////////////////////////////////////////

  //////////////////////////////////////////////////////////
  //<editor-fold defaultstate="collapsed" desc="Feature Extraction">
  def toSentences(sentenceML: SentenceML): Seq[Sentence] = {
    toParagraphs().flatMap(paragraph => {
      paragraph.toSentences(sentenceML)
    })
  }

  def toTokens(sentenceML: SentenceML, tokenizer: TokenizerImp): Seq[String] = {
    toSentences(sentenceML).flatMap(sentence => sentence.tokens(tokenizer))
  }

  def toMorphs(sentenceML: SentenceML, tokenizer: TokenizerImp, disambiguator: MorphTagger): Seq[String] = {
    toSentences(sentenceML).flatMap(sentence => {
      val tokens = sentence.tokens(tokenizer)
      disambiguator.predict(tokens)
    })
  }

  def toPosses(sentenceML: SentenceML, tokenizerImp: TokenizerImp, posTagger: PosTaggerImp): Seq[String] = {
    toSentences(sentenceML).flatMap(sentence => {
      val tokens = sentence.tokens(tokenizerImp)
      posTagger.test(tokens)
    })
  }

  //</editor-fold>
  ///////////////////////////////////////////////////////////

  def addParagraph(paragraph: String): this.type = {
    paragraphs = paragraphs :+ paragraph
    this
  }


  override def toString = s"Document($docid, $doctype, $author)"

  def toXML(): String = {
    if (isTweet()) {
      val closeXML = "</RESULT>\n"
      val containerXML = "<RESULT TYPE=\"CONTAINER\" LABEL=\"TWEET\">\n"
      val authorXML = "<RESULT TYPE=\"TEXT\" LABEL=\"AUTHORNAME\">\n"
      val textXML = "<RESULT TYPE=\"TEXT\" LABEL=\"TWEETTEXT\">\n"
      containerXML +
        authorXML +
        author +
        closeXML +
        textXML +
        XMLParser.docToXML(text) +
        closeXML +
        closeXML
    }
    else if (isArticle()) {
      val rootXML = "<ROOT LABEL=\"" + Resources.DocumentModel.ARTICLEDOC + "\">\n"
      val rootCloseXML = "</ROOT>\n"
      val closeXML = "</RESULT>\n"
      val containerXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"CONTAINER\">"

      val authorXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"AUTHORNAME\">\n" + author + "\n" + closeXML
      val titleXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"ARTICLETITLE\">\n" + title + "\n" + closeXML
      val genreXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"GENRE\">\n" + genre + "\n" + closeXML
      var textXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"ARTICLETEXT\">\n"
      for (paragraph <- paragraphs) {
        val paragraphXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"ARTICLEPARAGRAPH\">\n" +
          paragraph + "\n" + closeXML

        textXML += paragraphXML

      }

      textXML += closeXML

      rootXML + containerXML +
        authorXML +
        titleXML +
        genreXML +
        textXML +
        closeXML +
        rootCloseXML

    }
    else if (isBlog()) {
      val rootXML = "<ROOT LABEL=\"" + DocumentModel.BLOGDOC + "\">\n"
      val rootCloseXML = "</ROOT>\n"
      val closeXML = "</RESULT>\n"
      val containerXML = "<RESULT TYPE=\"CONTAINER\" LABEL=\"ARTICLE\">"

      val authorXML = "<RESULT TYPE=\"TEXT\" LABEL=\"AUTHORNAME\">\n" + author + "\n" + closeXML
      val titleXML = "<RESULT TYPE=\"TEXT\" LABEL=\"ARTICLETITLE\">\n" + title + "\n" + closeXML
      val genreXML = "<RESULT TYPE=\"TEXT\" LABEL=\"GENRE\">\n" + genre + "\n" + closeXML
      var textXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"ARTICLETEXT\">\n"
      for (paragraph <- paragraphs) {
        val paragraphXML = "<RESULT TYPE=\"ARTICLE\" LABEL=\"ARTICLEPARAGRAPH\">\n" +
          paragraph + "\n" + closeXML

        textXML += paragraphXML

      }

      textXML += closeXML

      rootXML + containerXML +
        authorXML +
        titleXML +
        genreXML +
        textXML +
        closeXML +
        rootCloseXML

    }
    else {
      ""
    }
  }


}

